File: sample/src/main/java/com/polidea/rxandroidble2/sample/SampleApplication.java
package com.polidea.rxandroidble2.sample;

import android.app.Application;
import android.content.Context;

import android.util.Log;
import com.polidea.rxandroidble2.RxBleClient;
import com.polidea.rxandroidble2.exceptions.BleException;
import com.polidea.rxandroidble2.internal.RxBleLog;
import io.reactivex.exceptions.UndeliverableException;
import io.reactivex.plugins.RxJavaPlugins;

public class SampleApplication extends Application {

    private RxBleClient rxBleClient;

    /**
     * In practise you will use some kind of dependency injection pattern.
     */
    public static RxBleClient getRxBleClient(Context context) {
        SampleApplication application = (SampleApplication) context.getApplicationContext();
        return application.rxBleClient;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        rxBleClient = RxBleClient.create(this);
        RxBleClient.setLogLevel(RxBleLog.VERBOSE);
        RxJavaPlugins.setErrorHandler(throwable -> {
            if (throwable instanceof UndeliverableException && throwable.getCause() instanceof BleException) {
                Log.v("SampleApplication", "Suppressed UndeliverableException: " + throwable.toString());
                return; // ignore BleExceptions as they were surely delivered at least once
            }
            // add other custom handlers if needed
            throw new RuntimeException("Unexpected Throwable in RxJavaPlugins error handler", throwable);
        });
    }
}

File: sample/src/main/java/com/polidea/rxandroidble2/sample/SampleApplication.kt
package com.polidea.rxandroidble2.sample

import android.app.Application
import android.content.Context

import com.polidea.rxandroidble2.RxBleClient
import com.polidea.rxandroidble2.internal.RxBleLog

class SampleApplication : Application() {

    private var rxBleClient: RxBleClient? = null

    override fun onCreate() {
        super.onCreate()
        rxBleClient = RxBleClient.create(this)
        RxBleClient.setLogLevel(RxBleLog.VERBOSE)
    }

    companion object {

        /**
         * In practise you will use some kind of dependency injection pattern.
         */
        fun getRxBleClient(context: Context): RxBleClient? {
            val application = context.applicationContext as SampleApplication
            return application.rxBleClient
        }
    }
}
